1234 - Add Kaltura integration to H5P videos

zifgu opened the pull request:
## Changes
* Adds ability to upload video files in Interactive Video H5Ps to Kaltura.
    * Interactive Video H5Ps where the video was added via upload are shown in the Kaltura upload dashboard.
    * Video files inside Interactive Video H5Ps can be uploaded Kaltura. Upon successful upload, the H5P content will be updated and the file will be deleted.
* Minor edits to the Kaltura upload dashboard: now indicates the type of node (H5P or Video), and adds a note that only Video nodes can use the Kaltura media player (H5P nodes still use the H5P player).

## Screenshot
![image](https://user-images.githubusercontent.com/44929104/184790360-d6a42214-9230-44a9-bb09-d729ccada83e.png)
## Issue Linkage
Closes #1234 
## PR Dependency
Depends on: #1169 
## Automated Testing
N/A
## Testing
* TODO

legendword reviewed (commented):
H5P interactive video nodes now show up in the Kaltura tab of Tapestry Settings modal. Uploading the video works without errors. The corresponding H5P interactive video node can be viewed as normal and the video has a URL that is pointing to the Kaltura video. The local video file is removed from the H5P content folder. The "Switch to use Kaltura player" option indeed does not switch the player for H5P videos.

When an error occurs during the upload, using the "Clean Uploaded Videos" in Wordpress Settings > Tapestry successfully changes the status of the upload and deleted the local video files for both H5P and URL videos. The "Switch to use Kaltura player" in this page also works as expected: only URL videos are switched to the Kaltura player.

Just to clarify, we no longer give the user the stack trace (see screenshot below) when an error occurs during video upload, right? I remember this was implemented in some other PR, but I'm not so sure. Could you let me know if we have changed this some place? Thanks.

<img width="784" alt="image" src="https://user-images.githubusercontent.com/12668055/186033277-47b9f4e3-3317-4526-b1ad-4263a0302892.png">

And also, the above alert does not go away when "Clean Uploaded Videos" is performed from the Wordpress Settings > Tapestry page. It stays visible every time I open the Settings modal. Is this supposed to happen? If not we might want to fix it (and see if it also occurs on #1169).

Other than this, I believe everything works as expected and the code changes are great. I just left a couple refactoring comments on the code.

zifgu commented on utilities/class.tapestry-helpers.php:
```diff
@@ -402,25 +431,125 @@ public static function getVideosToUpload($tapestryPostId)
+     * @param string $mediaURL
+     * @return string
+     */
+    public static function getH5PIdFromMediaURL($mediaURL)
+    {
+        $urlParts = explode('&id=', $mediaURL);
+        return count($urlParts) >= 2 ? $urlParts[1] : null;
+    }
```
Good idea - thank you! I have made the change.

zifgu commented on utilities/class.tapestry-helpers.php:
```diff
@@ -402,25 +431,125 @@ public static function getVideosToUpload($tapestryPostId)
+     * Gets the video source (URL or path) from an H5P content, by the H5P ID
+     *
+     * @param string|int $h5pId
+     * @return string|null          Returns null if the H5P ID is invalid or the H5P an interactive video
```
Fixed!

zifgu commented on utilities/class.tapestry-helpers.php:
```diff
@@ -447,6 +576,10 @@ public static function checkVideoFileSize($node)
     private static function _realFileSize($path)
     {
+        if (!$path) {
+            return PHP_INT_MAX;
```
Yes, I included this check because I could think of some cases where the path object would be null (e.g. if `H5P_DEFINED` is false, or if a video node's mediaURL is not a local upload, which was added in 1213). But on second thought, you're right - this doesn't make a lot of sense. All calls to `checkVideoFileSize` are indeed guarded by `videoCanBeUploaded`, which makes these cases impossible. I'll write down a precondition to `checkVideoFileSize` instead.

I think you're also right about the second point. If we really need `_realFileSize` to handle a null path in the future, we might want to let the caller customize the return value when the path is null, since `PHP_INT_MAX` might not be right for all use cases.

zifgu commented:
> Just to clarify, we no longer give the user the stack trace (see screenshot below) when an error occurs during video upload, right? I remember this was implemented in some other PR, but I'm not so sure. Could you let me know if we have changed this some place? Thanks.

Yeah, we shouldn't be. I made the change in 1213 and not 1168, which is my bad. Just fixed this on 1168 so that all other branches can inherit the change.

> And also, the above alert does not go away when "Clean Uploaded Videos" is performed from the Wordpress Settings > Tapestry page. It stays visible every time I open the Settings modal. Is this supposed to happen? If not we might want to fix it (and see if it also occurs on #1169).

Right now, "Clean Uploaded Videos" does not modify the saved upload error, so this is expected. I definitely see how it can be unintuitive - for example, the error is still shown even if all videos are "complete". I wasn't sure if we wanted to clear it (since it lets the user know that the previous upload was interrupted), or if we should conditionally clear it (e.g. only if all videos are marked "Complete"?). Maybe we can discuss this tomorrow.

zifgu commented:
> And also, the above alert does not go away when "Clean Uploaded Videos" is performed from the Wordpress Settings > Tapestry page. It stays visible every time I open the Settings modal. Is this supposed to happen? If not we might want to fix it (and see if it also occurs on #1169).

Thanks again for pointing this out. Just fixed on #1169; we now clear the upload error when "Clean Uploaded Videos" is run, and allow the user to dismiss the error alert.

zifgu commented:
I think no new behaviours have been added to this PR, so please feel free to just confirm the refactoring changes.

Bitajkr commented:
For whatever reason, I am unable to press the Kaltura button under the video content type (see image below) 
<img width="985" alt="Screen Shot 2022-10-26 at 2 58 59 PM" src="https://user-images.githubusercontent.com/113054791/198146527-7ed03106-3ffd-43a4-9d0a-6bcb44538155.png">

wynnset commented:
Hi @Bitajkr which site are you testing on?

legendword commented:
Since this pull request has been tested and reviewed, I will just do some testing to make sure nothing new on master breaks the changes here. After which this PR should be ready to merge.

legendword reviewed (approved):
Did some code refactoring to fix a few small problems, and tested the functionality. I was able to upload the video in a H5P node to Kaltura without problems.

wynnset commented:
@legendword there seems to be a test failing. Could you please take a look?

wynnset commented:
@legendword please fix the merge conflicts.

legendword commented:
> @legendword there seems to be a test failing. Could you please take a look?

I think that test failure was a one-time thing. I ran the test locally by checking out that commit and the test passed. Also all tests passed for subsequent and previous commits. I don't think that failure is related to any code changes that were introduced in this PR.
